### PR TITLE
Sticky dialog

### PR DESCRIPTION
**Pull request by andrewjtait**
* Screen is no longer scrollable when a dialog is open.
* Dialog will attach to the bottom of the browser if it gets too tall, and it's content will become scrollable.
* If a dialog has a form, the form buttons will become sticky to the bottom of the dialog while the dialog is attached to the bottom of the browser.
* Dialog can now use a prop called `height`, allowing developers to specify a set height for the dialog (the dialog will still attach to the bottom of the browser if it is taller than the browser's height).

sticky-footer

### CODE REVIEW

File: src/components/textarea/textarea.scss [hunk 1]
```diff
@@ -2,7 +2,6 @@
 
 .carbon-textarea__input {
   height: auto;
-  resize: none;
```
Comment (andrewjtait) [inline src/components/textarea/textarea.scss]: revert

File: src/components/dialog/dialog.js [hunk 2]
```diff
@@ -150,6 +159,7 @@ class Dialog extends Modal {
    */
   window = () => {
     return Browser.getWindow();
+    console.log(this._dialog);
```
Comment (andrewjtait) [inline src/components/dialog/dialog.js]: remove

**Comment by andrewjtait** [comment]
So far this PR assumed that the sticky footer would be used with a dialog, but it looks as though it also needs to support within a regular page.

We could adapt this PR so that `Form` maintains the state of a sticky footer, and the dialog controls it being on/off.

File: src/components/form/definition.js [hunk 3]
```diff
@@ -67,7 +69,7 @@ let definition = new Definition('form', Form, {
   propValues: {
     activeInput: '',
     cancelText: '',
-    children: `<Textbox
+    children: `<Textarea
```
Comment (SeanArmstrong) [inline src/components/form/definition.js]: Do you want to revert this too?

File: src/components/dialog/dialog.js [hunk 4]
```diff
@@ -154,56 +169,87 @@ class Dialog extends Modal {
...
+    if (this._content) {
+      // apply height to content based on height of title
+      const titleHeight = this._title ? this._title.offsetHeight : '0';
+      this._content.style.height = `calc(100% - ${titleHeight}px)`;
+    }
+
     this._dialog.style.top = `${midPointY}px`;
     this._dialog.style.left = `${midPointX}px`;
+
+    if (animating === true) {
```
Comment (SeanArmstrong) [inline src/components/dialog/dialog.js]: do we need the `=== true`

File: src/components/dialog/dialog.js [hunk 5]
```diff
@@ -213,6 +259,11 @@ class Dialog extends Modal {
     if (!this.props.title) { return null; }
 
     let title = this.props.title;
+    const classes = classNames(
+          'carbon-dialog__title', {
+            'carbon-dialog__title--has-subheader': this.props.subtitle
+          }
+        );
```
Comment (SeanArmstrong) [inline src/components/dialog/dialog.js]: indentation looks off here

File: src/components/form/form.js [hunk 6]
```diff
@@ -601,11 +671,19 @@ class Form extends React.Component {
    */
   formFooter = () => {
     const save = this.props.showSummary ? this.saveButtonWithSummary() : this.saveButton();
+    let padding = this.props.stickyFooterPadding;
+
+    if (padding && !padding.match(/px$/)) {
+      padding = `${padding}px`;
```
Comment (SeanArmstrong) [inline src/components/form/form.js]: We usually define a group of preset values. Do you think we should match this here?

File: src/components/dialog/dialog.js (hunk 5 above)
Comment (SeanArmstrong) [inline src/components/dialog/dialog.js]: Thought linter would of caught this :(

File: src/components/dialog/dialog.js (hunk 4 above)
Comment (andrewjtait) [inline src/components/dialog/dialog.js]: I think it sometimes passes the event, which would be truthy, so i needed this check.

File: src/components/form/form.js (hunk 6 above)
Comment (andrewjtait) [inline src/components/form/form.js]: No this is so they can tweak the padding to match the UI in it's none-sticky state - that could be anything.

File: src/components/dialog/dialog.js (hunk 5 above)
Comment (andrewjtait) [inline src/components/dialog/dialog.js]: The linter _did_ catch this, which is why I've done it as I have. The linter doesn't cater for `const` being 5 characters, vs `let` or `var` being 3 :(

Comment (SeanArmstrong) [inline src/components/dialog/dialog.js]: Im still not convinced but there is no need for this to block the PR. If the linter is happy

Comment (andrewjtait) [inline src/components/dialog/dialog.js]: its a difference between

```
var mything = {
      something: 'blah'
    };
```

and 

```
const mything = {
      something: 'blah'
    };
```

**Comment by andrewjtait** [comment]
closing in favour of https://github.com/Sage/carbon/pull/1428

**Pull request closed without merging**